attacks/README.md

```diff
@@ -0,0 +1,122 @@
+# Attacks against Hohha Dynamic XOR
+
+Here will be a collection of examples demonstrating attacks against the Hohha
+Dynamic XOR encryption algorithm.
+
+## Adaptive CPA2 and CCA2
+
+Chosen plain text (CPA) and chosen cipher text (CCA) attacks are equivalent
+against this algorithm.  In other words, a method of attack using chosen plain
+text is just as effective with chosen cipher text.  In any of the oracles for
+CCA or CPA attacks, just change encryption `-e` to decryption `-d` or vice
+versa, and the attack will still succeed.
+
+For CPA or CCA attacks, the adversary is able to send plain text messages to be
+encrypted, and receive the encrypted result, or vice versa.  The adversary is
+not allowed to see the key or observe the operation of the algorithm.  The
+adversary may only observe the result.  The adversary is only allowed to use
+one operation, either encryption or decryption, not both.
+
+In CPA attacks, the adversary gets to choose the plain text.  It is possible
+that the adversary gets to choose the salt, too, but more likely the salt will
+not be under control of the adversary.  It is still interesting to consider the
+strength of the algorithm when the adversary can choose the salt.
+
+In CCA attacks, the adversary gets to choose the cipher text.  Since the salt
+and cipher text would be transferred in the same message, and the adversary can
+choose the cipher text, it is realistic to assume that the adversary can affect
+all parts of the message, including the salt.
+
+### Key Recovery: Key Length
+
+The key length is supposed to be secret information.  A CCA2 attack is
+provided which will recover the key length.
+
+```
+# choose a secret key
+vi oracle-length.sh
+
+# recover the key length with CCA2
+./solve-length.py
+```
+
+### Key Recovery: Key Jumps
+
+The key number of jumps is supposed to be secret information.  A CCA2 attack is
```

Comment:
This attack is actually CCA1 (batch), not CCA2(adaptive), because the next request to the oracle does not depend on any previous responses.  I could have generated all the requests up front, sent the batch to the oracle, and then solved for key jumps or key length without sending any more requests.  The script was written as CCA2, because it was just easier to write it that way.